yellow-paper/docs/addresses-and-keys/addresses-and-keys.md

```diff
@@ -0,0 +1,903 @@
+- An address (user identifier) can be derived deterministically, before deploying an account contract.
+
+### Encryption and decryption
+
+Definitions (from the point of view of a user ("yourself")):
+- Incoming data: Data which has been created by someone else, and sent to yourself.
+- Outgoing data: Data which has been sent to somebody else, from you.
+- Internal Incoming data: Data which has been created by you, and has been sent to yourself.
+  - Note: this was an important observation by ZCash. Before this distinction, whenever a 'change' note was being created, it was being broadcast as incoming data, but that allowed a 3rd party who was only meant to have been granted access to view "incoming" data (and not "outgoing" data), was also able to learn that an "outgoing" transaction had taken place (including information about the notes which were spent). The addition of "internal incoming" keys enables a user to keep interactions with themselves private and separate from interactions with others.
+
+Requirements:
+- A user can non-interactively encrypt state changes, actions, and messages to some other user.
+- A user can encrypt a record of any actions, state changes, or messages, to _themselves_, so that they may re-sync their entire history of actions from their `seed`.
+- A user can derive siloed keys-used-for-encryption on a _per app_ basis.
+  - Reason: if one app-siloed key is compromised, it doesn't leak the private transaction data of other apps.
+- If Bob's keys-used-for-encryption are leaked, it doesn't leak the details of Bob's interactions with Alice.
+  - Note: I'm not sure we can achieve this, given that Bob can encrypt outgoing data for himself. If Bob didn't encrypt data for himself, this would be achieved.
+
+
+### Auditability
+
+Some app developers will wish to give users the option of sharing private transaction details with a trusted 3rd party.
+
+Note: The block hashes tree will enable a user to prove many things about their historical transaction history, including historical encrypted event logs. This feature will open up exciting audit patterns, where a user will be able to provably respond to questions without necessarily revealing their private data. However, sometimes this might be an inefficient pattern; in particular when a user is asked to prove a negative statement (e.g. "prove that you've never owned a rock NFT"). Proving such negative statements might require the user to execute an enormous recursive function to iterate through the entire tx history of the network, for example: proving that, out of all the encrypted events that the user _can_ decrypt, none of them relate to ownership of a rock NFT. Given this (possibly huge) inefficiency, these key requirements include the more traditional ability to share certain keys with a trusted 3rd party.
```

Comment:
Doesn't this "prove that you've never owned a rock NFT" require that the auditor is also able to decrypt tags in addition to note contents? (maybe that's ok, I haven't reached tag derivation yet)

Reply:
Well, the point I'm making in this example is that the auditor doesn't technically need any decryption keys: the user can in theory generate a gigantic proof (which might take years to execute) to prove this fact without giving the auditor any information at all. Which is amazing.

But if we consider the route of the user _optionally_ providing an auditor with shareable viewing keys, then there are two ways the auditor can decrypt: the fast way, or the brute-force way. Assuming auditors are running beefy machines, they can feasibly trial-decrypt all ciphertexts (ignoring tags altogether) to decrypt a user's notes. But yeah, the tag scheme allows tags to be derived using the shareable viewing keys, as well, so an auditor could be given the ability to derive any app-specific tag sequences, which would make their job of note identification easier.

Reply:
> will tag sequences be app-specific or user-specific

This will depend on the age-old question of whether they'll be constrained in the app or in the account contract.

The draft scheme allows Alice and Bob to generate a _master_ seuqence of tags (that they can use across all apps), or a sequence of tags _per app_. Of course, if it's up to the app to decide, the user doesn't get to choose whether they can use master or app-siloed tags; it'll be chosen for them (although an advanced app might still give them a choice, idk).

The shared secret (that forms the basis of the tag sequence) can be shared using the recipient's shareable incoming viewing public key, so an auditor with that key would have enough info to derive the tag sequence.